Add tom 256

ineiti opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before hitting that submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

Add the Tom-256 curve from https://eprint.iacr.org/2021/1183 and described in https://neuromancer.sk/std/other/Tom-256

closes: #923 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

z-tech commented on devbox.json:
Should this be here?

z-tech commented:
Yeah I see, I think that must_use has to be moved to the struct from the trait. Have to look a bit more.

z-tech commented:
@ineiti the error you see is fixed in this PR: https://github.com/arkworks-rs/algebra/pull/946

ineiti commented:
Wrt `devbox.*` - I like devbox to have a "out of the box shell environment". It uses `nix` to provide a development shell based on changing the PATH. But it should probably be in a separate PR, I'll remove it.

I squashed the commits and removed devbox. I also rebased on your PR.

ineiti commented:
@Pratyush why did you close this PR? I made it depend on the `z-tech/fix-werror-on-must-use-annotation` to get the CI running, but it is an independent PR.

Pratyush commented:
Hm I didn't close it. I think GitHub automatically closes PRs which don't have a base. Could you rebase to master? Thank you, and sorry for the trouble!

Pratyush reviewed (commented):
Left some small tweaks.
